File: sdk/borsh-incremental/go/reader.go
// Package borshincremental provides cursor-based reading of Borsh-serialized
// data with backward-compatible incremental deserialization.
//
// The key invariant: if the reader offset hasn't advanced when a read fails,
// the data is simply missing (trailing field) and the TryRead* methods return
// a default. If the offset advanced but the read still failed, the data is
// corrupt and the Read* methods return an error.
package borshincremental

import (
	"encoding/binary"
	"fmt"
	"math"
)

// Reader provides cursor-based reading of Borsh-serialized binary data.
type Reader struct {
	data   []byte
	offset int
}

// NewReader creates a new Reader over the given byte slice.
func NewReader(data []byte) *Reader {
	return &Reader{data: data, offset: 0}
}

// Offset returns the current read position.
func (r *Reader) Offset() int {
	return r.offset
}

// Remaining returns the number of unread bytes.
func (r *Reader) Remaining() int {
	return len(r.data) - r.offset
}

// --- Strict read methods (error on insufficient data) ---

func (r *Reader) ReadU8() (uint8, error) {
	if r.offset+1 > len(r.data) {
		return 0, fmt.Errorf("borsh: not enough data for u8 at offset %d", r.offset)
	}
	val := r.data[r.offset]
	r.offset++
	return val, nil
}

func (r *Reader) ReadBool() (bool, error) {
	v, err := r.ReadU8()
	return v != 0, err
}

func (r *Reader) ReadU16() (uint16, error) {
	if r.offset+2 > len(r.data) {
		return 0, fmt.Errorf("borsh: not enough data for u16 at offset %d", r.offset)
	}
	val := binary.LittleEndian.Uint16(r.data[r.offset:])
	r.offset += 2
	return val, nil
}

func (r *Reader) ReadU32() (uint32, error) {
	if r.offset+4 > len(r.data) {
		return 0, fmt.Errorf("borsh: not enough data for u32 at offset %d", r.offset)
	}
	val := binary.LittleEndian.Uint32(r.data[r.offset:])
	r.offset += 4
	return val, nil
}

func (r *Reader) ReadU64() (uint64, error) {
	if r.offset+8 > len(r.data) {
		return 0, fmt.Errorf("borsh: not enough data for u64 at offset %d", r.offset)
	}
	val := binary.LittleEndian.Uint64(r.data[r.offset:])
	r.offset += 8
	return val, nil
}

func (r *Reader) ReadU128() ([16]byte, error) {
	if r.offset+16 > len(r.data) {
		return [16]byte{}, fmt.Errorf("borsh: not enough data for u128 at offset %d", r.offset)
	}
	var val [16]byte
	copy(val[:], r.data[r.offset:r.offset+16])
	r.offset += 16
	return val, nil
}

func (r *Reader) ReadF64() (float64, error) {
	v, err := r.ReadU64()
	return math.Float64frombits(v), err
}

func (r *Reader) ReadPubkey() ([32]byte, error) {
	if r.offset+32 > len(r.data) {
		return [32]byte{}, fmt.Errorf("borsh: not enough data for pubkey at offset %d", r.offset)
	}
	val := [32]byte(r.data[r.offset : r.offset+32])
	r.offset += 32
	return val, nil
}

func (r *Reader) ReadIPv4() ([4]byte, error) {
	if r.offset+4 > len(r.data) {
		return [4]byte{}, fmt.Errorf("borsh: not enough data for ipv4 at offset %d", r.offset)
	}
	val := [4]byte(r.data[r.offset : r.offset+4])
	r.offset += 4
	return val, nil
}

func (r *Reader) ReadNetworkV4() ([5]byte, error) {
	if r.offset+5 > len(r.data) {
		return [5]byte{}, fmt.Errorf("borsh: not enough data for network_v4 at offset %d", r.offset)
	}
	val := [5]byte(r.data[r.offset : r.offset+5])
	r.offset += 5
	return val, nil
}

func (r *Reader) ReadString() (string, error) {
	length, err := r.ReadU32()
	if err != nil {
		return "", err
	}
	if length == 0 {
		return "", nil
	}
	if r.offset+int(length) > len(r.data) {
		return "", fmt.Errorf("borsh: not enough data for string of length %d at offset %d", length, r.offset)
	}
	val := string(r.data[r.offset : r.offset+int(length)])
	r.offset += int(length)
	return val, nil
}

func (r *Reader) ReadBytes(n int) ([]byte, error) {
	if r.offset+n > len(r.data) {
		return nil, fmt.Errorf("borsh: not enough data for %d bytes at offset %d", n, r.offset)
	}
	val := make([]byte, n)
	copy(val, r.data[r.offset:r.offset+n])
	r.offset += n
	return val, nil
}

func (r *Reader) ReadPubkeySlice() ([][32]byte, error) {
	length, err := r.ReadU32()
	if err != nil {
		return nil, err
	}
	if length == 0 {
		return nil, nil
	}
	if int(length)*32 > r.Remaining() {
		return nil, fmt.Errorf("borsh: not enough data for %d pubkeys at offset %d", length, r.offset)
	}
	result := make([][32]byte, length)
	for i := range int(length) {
		result[i], err = r.ReadPubkey()
		if err != nil {
			return nil, err
		}
	}
	return result, nil
}

func (r *Reader) ReadNetworkV4Slice() ([][5]byte, error) {
	length, err := r.ReadU32()
	if err != nil {
		return nil, err
	}
	if length == 0 {
		return nil, nil
	}
	if int(length)*5 > r.Remaining() {
		return nil, fmt.Errorf("borsh: not enough data for %d network_v4 at offset %d", length, r.offset)
	}
	result := make([][5]byte, length)
	for i := range int(length) {
		result[i], err = r.ReadNetworkV4()
		if err != nil {
			return nil, err
		}
	}
	return result, nil
}

func (r *Reader) ReadU32Slice() ([]uint32, error) {
	length, err := r.ReadU32()
	if err != nil {
		return nil, err
	}
	if length == 0 {
		return nil, nil
	}
	if int(length)*4 > r.Remaining() {
		return nil, fmt.Errorf("borsh: not enough data for %d u32s at offset %d", length, r.offset)
	}
	result := make([]uint32, length)
	for i := range int(length) {
		result[i], err = r.ReadU32()
		if err != nil {
			return nil, err
		}
	}
	return result, nil
}

// --- Try variants (return default when no bytes available at field boundary) ---

func (r *Reader) TryReadU8(def uint8) uint8 {
	if r.Remaining() < 1 {
		return def
	}
	v, _ := r.ReadU8()
	return v
}

func (r *Reader) TryReadBool(def bool) bool {
	if r.Remaining() < 1 {
		return def
	}
	v, _ := r.ReadBool()
	return v
}

func (r *Reader) TryReadU16(def uint16) uint16 {
	if r.Remaining() < 2 {
		return def
	}
	v, _ := r.ReadU16()
	return v
}

func (r *Reader) TryReadU32(def uint32) uint32 {
	if r.Remaining() < 4 {
		return def
	}
	v, _ := r.ReadU32()
	return v
}

func (r *Reader) TryReadU64(def uint64) uint64 {
	if r.Remaining() < 8 {
		return def
	}
	v, _ := r.ReadU64()
	return v
}

func (r *Reader) TryReadU128(def [16]byte) [16]byte {
	if r.Remaining() < 16 {
		return def
	}
	v, _ := r.ReadU128()
	return v
}

func (r *Reader) TryReadF64(def float64) float64 {
	if r.Remaining() < 8 {
		return def
	}
	v, _ := r.ReadF64()
	return v
}

func (r *Reader) TryReadPubkey(def [32]byte) [32]byte {
	if r.Remaining() < 32 {
		return def
	}
	v, _ := r.ReadPubkey()
	return v
}

func (r *Reader) TryReadIPv4(def [4]byte) [4]byte {
	if r.Remaining() < 4 {
		return def
	}
	v, _ := r.ReadIPv4()
	return v
}

func (r *Reader) TryReadNetworkV4(def [5]byte) [5]byte {
	if r.Remaining() < 5 {
		return def
	}
	v, _ := r.ReadNetworkV4()
	return v
}

func (r *Reader) TryReadString(def string) string {
	if r.Remaining() < 4 {
		return def
	}
	v, _ := r.ReadString()
	return v
}

func (r *Reader) TryReadPubkeySlice(def [][32]byte) [][32]byte {
	if r.Remaining() < 4 {
		return def
	}
	v, _ := r.ReadPubkeySlice()
	return v
}

func (r *Reader) TryReadNetworkV4Slice(def [][5]byte) [][5]byte {
	if r.Remaining() < 4 {
		return def
	}
	v, _ := r.ReadNetworkV4Slice()
	return v
}

func (r *Reader) TryReadU32Slice(def []uint32) []uint32 {
	if r.Remaining() < 4 {
		return def
	}
	v, _ := r.ReadU32Slice()
	return v
}

File: sdk/borsh-incremental/python/borsh_incremental/reader.py
"""Borsh incremental deserialization reader.

Provides cursor-based reading of Borsh-serialized binary data with
backward-compatible trailing field support via try_read_* methods.
"""

from __future__ import annotations

import struct


class IncrementalReader:
    """Cursor-based Borsh binary reader with incremental deserialization."""

    def __init__(self, data: bytes) -> None:
        self._data = data
        self._offset = 0

    @property
    def offset(self) -> int:
        return self._offset

    @property
    def remaining(self) -> int:
        return len(self._data) - self._offset

    # --- Strict read methods (raise on insufficient data) ---

    def read_u8(self) -> int:
        if self._offset + 1 > len(self._data):
            raise ValueError(f"borsh: not enough data for u8 at offset {self._offset}")
        v = self._data[self._offset]
        self._offset += 1
        return v

    def read_bool(self) -> bool:
        return self.read_u8() != 0

    def read_u16(self) -> int:
        if self._offset + 2 > len(self._data):
            raise ValueError(f"borsh: not enough data for u16 at offset {self._offset}")
        (v,) = struct.unpack_from("<H", self._data, self._offset)
        self._offset += 2
        return v

    def read_u32(self) -> int:
        if self._offset + 4 > len(self._data):
            raise ValueError(f"borsh: not enough data for u32 at offset {self._offset}")
        (v,) = struct.unpack_from("<I", self._data, self._offset)
        self._offset += 4
        return v

    def read_u64(self) -> int:
        if self._offset + 8 > len(self._data):
            raise ValueError(f"borsh: not enough data for u64 at offset {self._offset}")
        (v,) = struct.unpack_from("<Q", self._data, self._offset)
        self._offset += 8
        return v

    def read_u128(self) -> int:
        if self._offset + 16 > len(self._data):
            raise ValueError(f"borsh: not enough data for u128 at offset {self._offset}")
        low, high = struct.unpack_from("<QQ", self._data, self._offset)
        self._offset += 16
        return low | (high << 64)

    def read_f64(self) -> float:
        if self._offset + 8 > len(self._data):
            raise ValueError(f"borsh: not enough data for f64 at offset {self._offset}")
        (v,) = struct.unpack_from("<d", self._data, self._offset)
        self._offset += 8
        return v

    def read_bytes(self, n: int) -> bytes:
        if self._offset + n > len(self._data):
            raise ValueError(
                f"borsh: not enough data for {n} bytes at offset {self._offset}"
            )
        v = bytes(self._data[self._offset : self._offset + n])
        self._offset += n
        return v

    def read_pubkey_raw(self) -> bytes:
        """Read a 32-byte public key as raw bytes."""
        return self.read_bytes(32)

    def read_ipv4(self) -> bytes:
        return self.read_bytes(4)

    def read_network_v4(self) -> bytes:
        return self.read_bytes(5)

    def read_string(self) -> str:
        length = self.read_u32()
        if length == 0:
            return ""
        if self._offset + length > len(self._data):
            raise ValueError(
                f"borsh: not enough data for string of length {length} at offset {self._offset}"
            )
        s = self._data[self._offset : self._offset + length].decode("utf-8")
        self._offset += length
        return s

    def read_pubkey_raw_vec(self) -> list[bytes]:
        length = self.read_u32()
        return [self.read_pubkey_raw() for _ in range(length)]

    def read_network_v4_vec(self) -> list[bytes]:
        length = self.read_u32()
        return [self.read_network_v4() for _ in range(length)]

    # --- Try variants (return default when no bytes available) ---

    def try_read_u8(self, default: int = 0) -> int:
        if self.remaining < 1:
            return default
        return self.read_u8()

    def try_read_bool(self, default: bool = False) -> bool:
        if self.remaining < 1:
            return default
        return self.read_bool()

    def try_read_u16(self, default: int = 0) -> int:
        if self.remaining < 2:
            return default
        return self.read_u16()

    def try_read_u32(self, default: int = 0) -> int:
        if self.remaining < 4:
            return default
        return self.read_u32()

    def try_read_u64(self, default: int = 0) -> int:
        if self.remaining < 8:
            return default
        return self.read_u64()

    def try_read_u128(self, default: int = 0) -> int:
        if self.remaining < 16:
            return default
        return self.read_u128()

    def try_read_f64(self, default: float = 0.0) -> float:
        if self.remaining < 8:
            return default
        return self.read_f64()

    def try_read_pubkey_raw(self, default: bytes = b"\x00" * 32) -> bytes:
        if self.remaining < 32:
            return default
        return self.read_pubkey_raw()

    def try_read_ipv4(self, default: bytes = b"\x00" * 4) -> bytes:
        if self.remaining < 4:
            return default
        return self.read_ipv4()

    def try_read_network_v4(self, default: bytes = b"\x00" * 5) -> bytes:
        if self.remaining < 5:
            return default
        return self.read_network_v4()

    def try_read_string(self, default: str = "") -> str:
        if self.remaining < 4:
            return default
        return self.read_string()

    def try_read_pubkey_raw_vec(self, default: list[bytes] | None = None) -> list[bytes]:
        if self.remaining < 4:
            return default if default is not None else []
        return self.read_pubkey_raw_vec()

    def try_read_network_v4_vec(self, default: list[bytes] | None = None) -> list[bytes]:
        if self.remaining < 4:
            return default if default is not None else []
        return self.read_network_v4_vec()


class DefensiveReader:
    """Wrapper around IncrementalReader that uses try_read for all operations.

    All read methods return zero/empty defaults on insufficient data, matching
    Go's ByteReader behavior. This makes deserialization resilient to schema
    changes where new fields are added to the end of structs.
    """

    def __init__(self, data: bytes) -> None:
        self._r = IncrementalReader(data)

    @property
    def offset(self) -> int:
        return self._r.offset

    @property
    def remaining(self) -> int:
        return self._r.remaining

    def read_u8(self) -> int:
        return self._r.try_read_u8(0)

    def read_bool(self) -> bool:
        return self._r.try_read_bool(False)

    def read_u16(self) -> int:
        return self._r.try_read_u16(0)

    def read_u32(self) -> int:
        return self._r.try_read_u32(0)

    def read_u64(self) -> int:
        return self._r.try_read_u64(0)

    def read_u128(self) -> int:
        return self._r.try_read_u128(0)

    def read_f64(self) -> float:
        return self._r.try_read_f64(0.0)

    def read_pubkey_raw(self) -> bytes:
        return self._r.try_read_pubkey_raw(b"\x00" * 32)

    def read_ipv4(self) -> bytes:
        return self._r.try_read_ipv4(b"\x00" * 4)

    def read_network_v4(self) -> bytes:
        return self._r.try_read_network_v4(b"\x00" * 5)

    def read_string(self) -> str:
        return self._r.try_read_string("")

    def read_pubkey_raw_vec(self) -> list[bytes]:
        return self._r.try_read_pubkey_raw_vec([])

    def read_network_v4_vec(self) -> list[bytes]:
        return self._r.try_read_network_v4_vec([])

    def read_bytes(self, n: int) -> bytes:
        """Read n bytes, returning zero bytes if insufficient data."""
        if self._r.remaining < n:
            return b"\x00" * n
        return self._r.read_bytes(n)
